src/globus_sdk/transport/requests.py

```diff
@@ -0,0 +1,156 @@
+
+import requests
+
+from globus_sdk import config, exc
+from globus_sdk.authorizers import GlobusAuthorizer
+from globus_sdk.transport.encoders import (
+    FormRequestEncoder,
+    JSONRequestEncoder,
+    RequestEncoder,
+)
+from globus_sdk.transport.retry import RetryContext, RetryPolicy
+from globus_sdk.version import __version__
+
+
+class RequestsTransport:
+    """
+    The RequestsTransport handles HTTP request sending and retries.
+
+    It receives raw request information from a client class, and then performs the
+    following steps
+    - encode the data in a prepared request
+    - repeatedly send the request until no retry is requested by the retry policy
+    - return the last response or reraise the last exception
+
+    :param verify_ssl: Explicitly enable or disable SSL verification. This parameter
+        defaults to True, but can be set via the ``GLOBUS_SDK_VERIFY_SSL`` environment
+        variable. Any non-``None`` setting via this parameter takes precedence over the
+        environment variable.
+    :type verify_ssl: bool
+    :param http_timeout: Explicitly set an HTTP timeout value in seconds. This parameter
+        defaults to 60s but can be set via the ``GLOBUS_SDK_HTTP_TIMEOUT`` environment
+        variable. Any value set via this parameter takes precedence over the environment
+        variable.
+    :type http_timeout: int
+    """
+
+    # the retry policy may enforce a lower maximum, but this sets a hard cap to avoid
+    # potential infinite loop with a bad policy
+    max_retries: int = 20
+
+    #: the encoders are a mapping of encoding names to encoder objects
+    encoders: typing.Dict[str, RequestEncoder] = {
+        "text": RequestEncoder(),
+        "json": JSONRequestEncoder(),
+        "form": FormRequestEncoder(),
+    }
+
+    BASE_USER_AGENT = f"globus-sdk-py-{__version__}"
+
+    def __init__(
+        self,
+        verify_ssl: typing.Optional[bool] = None,
+        http_timeout: typing.Optional[float] = None,
+        retry_policy: typing.Optional[RetryPolicy] = None,
+    ):
+        self.session = requests.Session()
+        self.verify_ssl = config.get_ssl_verify(verify_ssl)
+        self.http_timeout = config.get_http_timeout(http_timeout)
+        self.retry_policy = retry_policy if retry_policy else RetryPolicy()
+        self._user_agent = self.BASE_USER_AGENT
+
+    @property
+    def user_agent(self):
+        return self._user_agent
+
+    @user_agent.setter
+    def user_agent(self, value):
+        self._user_agent = f"{self.BASE_USER_AGENT}/{value}"
+
+    @property
+    def _headers(self):
+        return {"Accept": "application/json", "User-Agent": self.user_agent}
+
+    def _encode(
+        self,
+        method: str,
+        url: str,
+        params: typing.Optional[typing.Dict] = None,
+        data: typing.Union[typing.Dict, typing.List, str, None] = None,
+        headers: typing.Optional[typing.Dict[str, str]] = None,
+        encoding: typing.Optional[str] = None,
```

Comment:
Let's punt on this.

I like strings for this because it makes it possible to add a custom encoder. Imagine
```python
def add_validating_encoder(name: str, validator: Callable):
    class ValidatingEncoder(JSONRequestEncoder):
        def encode(self, method, url, params, data, headers):
            validator(data)
            return super().encode(method, url, params, data, headers)
    RequestsTransport.encoders[name] = ValidatingEncoder()
```

Maybe I'm speculating too much here and nobody would do this. We can switch to an enum, but let's consider that after we get the rest of this in, yeah?